src/System.Net.HttpListener/src/System/Net/Managed/HttpListenerRequest.Managed.cs

```diff
@@ -383,32 +380,46 @@ public bool KeepAlive
 
         public Guid RequestTraceIdentifier => Guid.Empty;
 
-        public IAsyncResult BeginGetClientCertificate(AsyncCallback requestCallback, object state)
+        private IAsyncResult BeginGetClientCertificateCore(AsyncCallback requestCallback, object state)
         {
-            Task<X509Certificate2> getClientCertificate = new Task<X509Certificate2>(() => GetClientCertificate());
-            return TaskToApm.Begin(getClientCertificate, requestCallback, state);
+            var asyncResult = new GetClientCertificateAsyncResult(this, state, requestCallback);
+
+            // The certificate is already retrieved by the time this method is called.
+            ClientCertState = ListenerClientCertState.Completed;
+            asyncResult.InvokeCallback(GetClientCertificate());
```

Comment:
The reason I went for this is because `GetClientCertificate` evaluates to a simple field access:

See: https://github.com/dotnet/corefx/blob/d375d3a0d90fd594c2995d5dbd5126ff3a88260f/src/System.Net.HttpListener/src/System/Net/Managed/HttpListenerRequest.Managed.cs#L368
```cs
public X509Certificate2 GetClientCertificate() => _context.Connection.ClientCertificate;
```

> Task.Run, Task.Factory.StartNew, or the ctor but with a subsequent call to .Start).

Which do you prefer? Why are there so many choices?!